ts/tree/leaves/Leaf.ts

```diff
@@ -0,0 +1,49 @@
+import { LevelUp } from "levelup";
+import { Hashable } from "../../interfaces";
+
+export type LeafFactoryFunc<T extends Leaf<Hashable>> = (
+    itemId: number,
+    itemHash: string
+) => Promise<T>;
+
+export const getLeafKey = (
+    name: string,
+    itemID: number,
+    itemHash: string
+): string => `${name}${itemID}${itemHash}`;
+
+export abstract class Leaf<Item extends Hashable> {
+    abstract readonly name: string;
+    abstract readonly db: LevelUp;
+
+    public readonly item: Item;
+    public readonly itemID: number;
+
+    constructor(item: Item, itemID: number) {
+        this.item = item;
+        this.itemID = itemID;
+    }
+
+    getKey() {
+        return getLeafKey(this.name, this.itemID, this.item.hash());
+    }
+
+    get key() {
+        return this.getKey();
+    }
+
+    serialize(): string {
+        return this.item.encode();
+    }
+
+    abstract deserialize(bytes: string): Item;
```

Comment:
since both Pubkey and State has a `fromEncoded` method, should we directly implement this method here or keep it abstract and implement later?

Reply:
I would just get rid of it here and it's implementations as well since we are using the static factory method with db.